content/posts/2019-08-03---Going-from-ember-to-vue.md

```diff
@@ -0,0 +1,284 @@
+}
+</script>
+```
+
+Ok, it might seem like there is a lot there, but not every component will need all of these things.
+I just wanted some examples of the major key pieces.
+
+First off, all of the markup is wrapped in a `template`.
+After that, you are writing normal HTML.
+You call other components just like `<AngleBracket />` components in Ember, but the differences (for now) is that you have to import them.
+This saves from potential name collision, as an npm package might have a component with the same name as one of yours.
+You could do `import SomeCustomComponent as Monkey from '@something/components';` and rename a component without any magic.
+
+## Methods/Actions
+
+When calling a component, we have similar patterns as Ember.
+You need to append a `:` to properties you pass down to a component and `@` for events that map to functions (think `actions`).
+`methods` are the equivalent of `actions` in Ember -- if there is a function that is going to be called from the template, it is in the `methods` hash.
+
+If you look at the `someMethod` function, you will see the last line calls `this.emit('something-that-bubbles');`.
+This is how we accomplish something similar to "Data Down, Actions Up" (DDAU).
+It's a bit different and shouldn't be used for every case, but basically, when we `emit` something, any parent component can listen to that event (in this case `something-that-bubbles`) and respond to it.
+This means that child component does not need to receive an action as a parameter.
+All it does is emit something, pass any properties along with it and then it stops caring.
+It's up to the parent component to listen to it and respond.
+
+One of the things I disliked with the actions helper in Ember was when the child component could call `this.passedAction()` and I would get an error that it wasn't a function because I forgot to pass it down, the child component stopped working.
+With emitting, the child component can continue to function.
+To be clear: making API calls to your backend should happen through the Vuex store -- mutating applicaton state should not be done with `emit`.
+However, if you have buttons component that toggles something on and off, this is a good case for using `emit` as the parent component is most likely the only thing that cares about the event.
+
+## Computed Properties
+
+Properties that needs to be available on the template, need to be declared in the hash returned by the `data` function.
+This ends up being really handy in documenting what shows up in your template.
+Computed properties are declared in the `computed` hash and behave very similarly.
+
+I love computed properties in Ember and was a bit miffed when there wasn't something similar in React.
+With Vue, you declare your computed properties as functions and Vue is smart enough to observe anything you declare within the function.
+So, we go from something like this Ember:
+
+```javascript
+fullName: computed('firstName', 'lastName', function() {
```

Comment:
or:
```ts
@computed('firstName', 'lastName')
get fullName() {
  return `${this.firstName} ${this.lastName`;
}
```
in Ember 3.10+

Or in Ember Octane
```ts
get fullName() {
  return `${this.firstName} ${this.lastName`;
}
```

Reply:
Dang, I've only been out of the ember game for half a year, and I feel like so much has so much more syntactic sugar 😅

Reply:
it's not syntactic sugar, if it's native JS :)

Reply:
Last I checked, decorators were only at stage-2 for tc39 proposals: https://github.com/tc39/proposals#stage-2

Reply:
yeah, but all the major ecosystems (except React) use them, they are for sure happening, but all the major players have found them so useful, that they're happening no matter what. The final details are being worked out on implementation specifics. A couple TC39 meetings ago, the Node team brought up some performance concerns with the stage2 proposal's dynamic nature of decorators, so the next iteration is going to be all static.

Reply:
The recommendation of the core team is to use `.extend({` with classic components, so the code in the article is the recommended one.

Reply:
> yeah, but all the major ecosystems (except React) use them

Actually Vue is not going to use them - they are moving into function-based component API (https://github.com/vuejs/rfcs/pull/42)

Reply:
I'm pretty excited about function components, despite the backlash from the initial proposal.  I feel like they are so much easier to reason about.